components/util/Anchors.tsx

```diff
@@ -73,9 +87,62 @@ export function AnchorsSetter(props) {
 
   useEffect(() => {
     setAnchors(anchors);
+    setActiveAnchor(anchors[0]);
 
-    return () => setAnchors([]);
+    return () => {
+      setAnchors([]);
+      setActiveAnchor(null);
+    };
   }, [router?.pathname]);
 
+  useLayoutEffect(() => {
+    let raf: number;
+    const anchorEls = anchors.map((a: Anchor) => document.getElementById(a.id));
+
+    const handleScroll = () => {
+      raf = requestAnimationFrame(() => {
+        const anchorId = anchorEls.reduce(
+          (acc: { score: number; el: { id: string } }, next) => {
+            const { y } = next.getBoundingClientRect();
+            const nextScore = Math.abs(SCROLL_OFFSET - y);
+
+            if (!acc) {
+              return { el: next, score: nextScore };
+            }
+
+            return nextScore < acc.score ? { el: next, score: nextScore } : acc;
+          },
+          null
+        ).el.id;
+
+        if (anchorId !== activeAnchor?.id) {
+          setActiveAnchor(anchors.find((a: Anchor) => a.id === anchorId));
+        }
+      });
+    };
+
+    window.addEventListener("scroll", handleScroll);
+    window.addEventListener("resize", handleScroll);
+
+    return () => {
+      window.removeEventListener("scroll", handleScroll);
+      window.removeEventListener("resize", handleScroll);
+      cancelAnimationFrame(raf);
```

Comment:
This is the cleanup function that gets called when the component unmounts. Here we clean up our event handlers and `cancelAnimationFrame(raf)` so they don't get called anymore.